File: solo-leveling-player-service/src/main/java/com/sleepkqq/sololeveling/player/service/service/player/CountExperienceService.java
package com.sleepkqq.sololeveling.player.service.service.player;

import org.springframework.stereotype.Service;

@Service
public class CountExperienceService {

  private static final int FIRST_LEVEL_PLAYER_EXPERIENCE = 100;
  private static final int NEXT_LEVEL_PLAYER_EXPERIENCE = 10;

  private static final int PLAYER_TASK_TOPIC_COEFFICIENT = 5;
  private static final int FIRST_LEVEL_TOPIC_EXPERIENCE =
      FIRST_LEVEL_PLAYER_EXPERIENCE / PLAYER_TASK_TOPIC_COEFFICIENT;
  private static final int NEXT_LEVEL_TOPIC_EXPERIENCE =
      NEXT_LEVEL_PLAYER_EXPERIENCE / PLAYER_TASK_TOPIC_COEFFICIENT;

  public int countPlayerExperienceToNextLevel(int level) {
    return countExperienceToNextLevel(
        level, FIRST_LEVEL_PLAYER_EXPERIENCE, NEXT_LEVEL_PLAYER_EXPERIENCE
    );
  }

  public int countTopicExperienceToNextLevel(int level) {
    return countExperienceToNextLevel(
        level, FIRST_LEVEL_TOPIC_EXPERIENCE, NEXT_LEVEL_TOPIC_EXPERIENCE
    );
  }

  private int countExperienceToNextLevel(
      int level,
      int firstLevelExperience,
      int nextLevelExperience
  ) {
    return firstLevelExperience + (nextLevelExperience * (level - 1));
  }
}

File: solo-leveling-player-service/src/main/kotlin/com/sleepkqq/sololeveling/player/service/service/player/CountExperienceService.kt
package com.sleepkqq.sololeveling.player.service.service.player

import com.sleepkqq.sololeveling.player.model.entity.player.enums.LevelType
import org.springframework.stereotype.Service

@Service
class CountExperienceService {

	private companion object {
		const val FIRST_LEVEL_PLAYER_EXPERIENCE = 100
		const val NEXT_LEVEL_PLAYER_EXPERIENCE = 10

		const val PLAYER_TASK_TOPIC_COEFFICIENT = 5
		const val FIRST_LEVEL_TOPIC_EXPERIENCE =
			FIRST_LEVEL_PLAYER_EXPERIENCE / PLAYER_TASK_TOPIC_COEFFICIENT
		const val NEXT_LEVEL_TOPIC_EXPERIENCE =
			NEXT_LEVEL_PLAYER_EXPERIENCE / PLAYER_TASK_TOPIC_COEFFICIENT
	}

	fun countExperienceToNextLevel(level: Int, levelType: LevelType): Int =
		when (levelType) {
			LevelType.PLAYER -> countExperienceToNextLevel(
				level,
				FIRST_LEVEL_PLAYER_EXPERIENCE,
				NEXT_LEVEL_PLAYER_EXPERIENCE
			)

			LevelType.TASK_TOPIC -> countExperienceToNextLevel(
				level,
				FIRST_LEVEL_TOPIC_EXPERIENCE,
				NEXT_LEVEL_TOPIC_EXPERIENCE
			)
		}

	private fun countExperienceToNextLevel(
		level: Int,
		firstLevelExperience: Int,
		nextLevelExperience: Int
	): Int = firstLevelExperience + (nextLevelExperience * (level - 1))
}
